src/components/otpInput/OTPInput.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from "react";
+import { useEffect, useState } from "react";
+import { Pressable, TextInput } from "react-native";
+import Animated from "react-native-reanimated";
+import { IOStyles } from "../../core/IOStyles";
+import { LabelSmall } from "../typography";
+import { triggerHaptic } from "../../functions";
+import { VSpacer } from "../spacer";
+import { useErrorShakeAnimation } from "../../utils/hooks/useErrorShakeAnimation";
+import { BoxedInput } from "./BoxedInput";
+
+type Props = {
+  value: string;
+  onValueChange: (value: string) => void;
+  length: number;
+  secret?: boolean;
+  autocomplete?: boolean;
+  onValidate?: (value: string) => boolean;
+  errorMessage?: string;
+};
+
+/**
+ * `OTPInput` is a component that allows the user to enter a one-time password.
+ * It has an hidden `TextInput` that is used to handle the keyboard and the focus.
+ * The input handles the autocompletion of the OTP code.
+ * @param value - The value of the OTP code
+ * @param onValueChange - The function to call when the value changes
+ * @param length - The length of the OTP code
+ * @param secret - If the OTP code should be hidden
+ * @param autocomplete - If the OTP code should be autocompleted
+ * @param onValidate - The function to call when the OTP code is validated
+ * @param errorMessage - The error message to display
+ * @returns
+ */
+export const OTPInput = ({
+  value,
+  onValueChange,
+  length,
+  onValidate,
+  errorMessage = "",
+  secret = false,
+  autocomplete = false
+}: Props) => {
+  const [hasFocus, setHasFocus] = useState(false);
+  const [hasError, setHasError] = useState(false);
+  const [inputValue, setInputValue] = useState(value);
+
+  const { translate, animatedStyle, shakeAnimation } = useErrorShakeAnimation();
+
+  const inputRef = React.createRef<TextInput>();
+  const handleChange = (value: string) => {
+    if (value.length > length) {
+      return;
+    }
+    setInputValue(value);
+    onValueChange(value);
+  };
+
+  useEffect(() => {
+    if (onValidate && value.length === length) {
+      const isValid = onValidate(value);
+
+      if (!isValid) {
+        setHasError(true);
+        triggerHaptic("notificationError");
+
+        // eslint-disable-next-line functional/immutable-data
+        translate.value = shakeAnimation();
+
+        const timer = setTimeout(() => {
+          setHasError(false);
+          setInputValue("");
+          onValueChange("");
+        }, 500);
+        return () => clearTimeout(timer);
+      }
+    }
+    return;
+  }, [value, onValidate, length, onValueChange, translate, shakeAnimation]);
```

Comment:
```suggestion
const inputRef = useRef<TextInput>(null);

  const timerRef = useRef<NodeJS.Timeout>();
  const handleValidate = (val: string) => {
    if (!onValidate || val.length < length) {
      return;
    }

    const isValid = onValidate(val);

    if (!isValid) {
      setHasError(true);
      triggerHaptic("notificationError");

      // eslint-disable-next-line functional/immutable-data
      translate.value = shakeAnimation();

      if (timerRef.current) {
        clearTimeout(timerRef.current);
      }

      // eslint-disable-next-line functional/immutable-data
      timerRef.current = setTimeout(() => {
        setHasError(false);
        setInputValue("");
        onValueChange("");
      }, 500);
    }
  };
  const handleChange = (value: string) => {
    if (value.length > length) {
      return;
    }
    setInputValue(value);
    onValueChange(value);
    handleValidate(value);
  };

`

Reply:
I'm just missing one point on this suggestion, what would trigger the validation callback?

Reply:
It should be triggered by the handleChange function

Reply:
Tested out, it actually works perfectly!

Just some changes:
- I left the `createRef` for `inputRef` cause of type specifications
- I left the `useEffect` to handle the timeout clear when the component is unmounted

Reply:
[513c46a](https://github.com/pagopa/io-app-design-system/pull/161/commits/513c46afa9a0aff1ba79b7afdffaab70fa2b6946)